Improve code logic and add fetch genome by ensembl and assembly name

bilalebi opened the pull request:
### JIRA Ticket
https://www.ebi.ac.uk/panda/jira/browse/EA-1090

### Changes
Added the possibility to fetch genome by `ensembl_name` and `assembly_name` (`genebuild_id` will be added in the future)

---

Update:

> [Major changes](https://github.com/Ensembl/ensembl-metadata-api/pull/31#issuecomment-1651656468)
> * Added fetch genome by ensembl name and assembly name 
> * Remove mutual exclusivity in `fetch_genomes()` and `fetch_sequences()`
> * Improved `fetch_taxonomy_names()` to include `ncbi_common_name`, `alternative_names` and `common_name`
> * Improved `fetch_sequences()` to include `assembly_uuid`
> * Added `fetch_genome_by_keyword()`
> * Added and fixed broken tests
> * Improved the overall code logic
> * Added the possibility to fetch unreleased genomes
> * ...

This branch is also used to update the API to be used by [ensembl-metadata-service](https://github.com/Ensembl/ensembl-metadata-service)


### Related PR:
https://github.com/Ensembl/ensembl-metadata-service/pull/25

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -80,9 +80,46 @@ def fetch_taxonomy_ids(self, taxonomy_names):
-                      taxonomy_id=None, group=None, group_type=None, unreleased_only=False, site_name=None,
-                      release_type=None, release_version=None, current_only=True):
+    def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=None, assembly_name=None,
+                      ensembl_name=None, genebuild_id=None, taxonomy_id=None, group=None, group_type=None,
```
genebuild_id is not implemented and should not be part of this PR.
This could be it's own method within the class as it isn't trivial to get. It will also only ever return one genome so it could be interesting.
Better might be dataset_name or other aspect of dataset. Those could be very useful.

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -129,11 +169,18 @@ def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=Non
         elif assembly_accession is not None:
             genome_select = genome_select.filter(Assembly.accession.in_(assembly_accession))
 
+        # Handle when both assembly_name and ensembl_name are provided
+        # Todo: add genebuild_id once it's in the metadata DB (EA-1090)
+        if assembly_name is not None and ensembl_name is not None:
+            genome_select = genome_select.filter(
+                db.and_(Assembly.name.in_([assembly_name]), Organism.ensembl_name.in_([ensembl_name]))
+            )
+
```
This is against the logic of the method.
Each separate parameter adds an individual filter which are combined at the end. 
Checking two at once is excessive and unnecessary. 
Adding assembly name is a good idea, but I would put it on its own.

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -104,11 +141,12 @@ def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=Non
                 .join(OrganismGroupMember.organism_group) \
                 .filter(OrganismGroup.type.in_(group_type)).filter(OrganismGroup.name.in_(group))
 
+        # Apply additional filters based on the provided parameters
```
Have you tested these together? There are a few joins that might have to be done together before the filtering, but I am not sure.

bilalebi replied:
I added some tests and played with the code separately, I didn't get any unexpected behaviour

dpopleton commented on src/tests/test_api.py:
```diff
@@ -82,6 +82,12 @@ def test_fetch_genomes_by_genome_uuid(self, multi_dbs):
+        test = conn.fetch_genomes(assembly_name='WBcel235', ensembl_name='caenorhabditis_elegans')
+        assert test[0].Organism.scientific_name == 'Caenorhabditis elegans'
+
     def test_fetch_genomes_by_assembly_accession(self, multi_dbs):
```
Would you add a couple more tests to check your changes?

- everything is specified
- specifying with a few direct conflicts

bilalebi replied:
Done!

bilalebi commented:
Hi @vinay-ebi 
Please feel free to take a look at this PR, some of the major changes:
* Added fetch genome by ensembl name and assembly name 
* Remove mutual exclusivity in `fetch_genomes()` and `fetch_sequences()`
* Improved `fetch_taxonomy_names()` to include `ncbi_common_name`, `alternative_names` and `common_name`
* Added `fetch_genome_by_keyword()`
* Added and fixed broken tests
* Added `__repr__` functions in the models
* Improved the overall code logic

> NB: I commented 3 out of 4 tests in the updater because they were failing and they need to be fixed based on the changes @dpopleton made in this PR https://github.com/Ensembl/ensembl-metadata-api/pull/32 that can be done later

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -80,64 +112,118 @@ def fetch_taxonomy_ids(self, taxonomy_names):
+        # Apply additional filters based on the provided parameters
         if unreleased_only:
+            # Question (Bilal): When adding unreleased genomes do we populate genome_releases table?
+            # if that's the case the test data doesn't do so (this will return en empty list)
```
Unreleased genomes will not have a genome_release. Your logic is odd in my mind as I would look for genomes without a genome_release, but if it works I am fine with it.

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -80,64 +112,118 @@ def fetch_taxonomy_ids(self, taxonomy_names):
                 genome_select = genome_select.filter(GenomeRelease.is_current == 1)
 
-            if release_version is not None:
+            if release_version != 0.0:
```
So when we are releasaing a genome we will create the genome_release and link the genome_datasets. Until then there is nothing for those genomes.
The 0.0 here is not proper for any aspect of the API

bilalebi replied:
done

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -208,31 +294,94 @@ def fetch_genomes_by_scientific_name(
+            genome_query = genome_query.where(EnsemblRelease.version <= release_version)
+
+        with self.metadata_db.session_scope() as session:
+            session.expire_on_commit = False
```
Make sure you need the session to keep in memory here. If so, fine.

bilalebi replied:
Yes, it's needed.
When commenting it some tests start throwing this error:
```
sqlalchemy.orm.exc.DetachedInstanceError: Instance <DatasetType at 0x7fc5c05a13d0> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)
```

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -249,48 +398,106 @@ def fetch_sequences_by_assembly_accession(
 
             if unreleased_datasets:
                 genome_select = genome_select.filter(GenomeDataset.release_id.is_(None)) \
                     .filter(GenomeDataset.is_current == 0)
```
Don't Filter off of is_current for unreleased. Old datasets will be present. Filter off the absence of a release.

bilalebi replied:
fixed

dpopleton commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -249,48 +398,106 @@ def fetch_sequences_by_assembly_accession(
+                    .join(DatasetAttribute, DatasetAttribute.dataset_id == Dataset.dataset_id) \
+                    .join(Attribute, Attribute.attribute_id == DatasetAttribute.attribute_id)
+
+            if dataset_name is not None and "all" not in dataset_name:
```
Why is the "all" here?

bilalebi replied:
By default the API `dataset_name="assembly"`; `all` allows us to fetch everything
Please take a look at this discussion for more context: https://github.com/Ensembl/ensembl-metadata-api/commit/a8eb73d7e3d61d84510dc957cf87777f93e4434c#r121262595

dpopleton replied:
Ok, if it works, I am happy.

dpopleton commented on src/ensembl/production/metadata/api/models/genome.py:
```diff
@@ -36,6 +36,9 @@ class Genome(Base):
     # organism_id to organism
     organism = relationship("Organism", back_populates="genomes")
 
+    def __repr__(self):
```
If you are going to add the repr, please be standard between all model classes. These are of a different format than assembly.

bilalebi replied:
I removed them for now

dpopleton reviewed (changes requested):
So a few minor comments. It looks good for the most part. I have not tested it further or checked for conflicts , so I hope you have. :)

bilalebi commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -249,48 +398,106 @@ def fetch_sequences_by_assembly_accession(
                 genome_select = genome_select.filter(Dataset.dataset_uuid.in_(dataset_uuid))
 
             if unreleased_datasets:
                 genome_select = genome_select.filter(GenomeDataset.release_id.is_(None)) \
```
Done

bilalebi commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -201,31 +273,94 @@ def fetch_genomes_by_scientific_name(
 
-        seq_select = db.select(AssemblySequence, )
+        seq_select = db.select(
+            Genome, Assembly, AssemblySequence
```
Yes, they're used in the Metadata service to fetch data (e.g. `accession`, `level`...)

bilalebi commented on src/ensembl/production/metadata/api/genome.py:
```diff
@@ -73,66 +74,137 @@ def fetch_taxonomy_ids(self, taxonomy_names):
 
+        # Check if genome is released
+        with self.metadata_db.session_scope() as session:
+            session.expire_on_commit = False
```
Fixed
